keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -102,26 +142,77 @@ proposing the following changes.
+bogged down.
+
+TODO: ask antoine if these stories are complete/sufficient
+-->
+
+
+#### Story 1
+
+As a CRD owner, I should be able to author my CRD such that the openapi schema
+indicates which fields are members of this oneOf, so that when users crate
+custom resources, the oneOf fields will be automatically validated, without me
+having to write custom validation logic.
+
+#### Story 2
+
+As a maintainer of existing builtin APIs, I should be able to add new fields to
+a union and have it behave as expected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+TODO: ? 
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+- We need to ensure we do not break existing union types. This can be done by
+  not forcing existing unions to conform to the newly proposed union semantics.
+  Integration testing with older types should give us the confidence to be sure
+  we have done so.
+- There is a lot of risk for errors when there exists skew between clients and
+  server. In the section on normalization, we discuss mitigating these risks.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
 ### Go tags
 
 We're proposing a new type of tags for go types (in-tree types, and also
 kubebuilder types):
 
-- `// +union` before a structure means that the structure is a union. All the
-  fields must be optional (beside the discriminator) and will be included as
-  members of the union. That structure CAN be embedded in another structure.
-- `// +unionDeprecated` before a field means that it is part of the
-  union. Multiple fields can have this prefix. These fields MUST BE optional,
-  omitempty and pointers. The field is named deprecated because we don't want
-  people to embed their unions directly in structures, and only exist because of
-  some existing core types (e.g. `Value` and `ValueFrom` in
-  [EnvVar](https://github.com/kubernetes/kubernetes/blob/3ebb8ddd8a21b/staging/src/k8s.io/api/core/v1/types.go#L1817-L1836)).
 - `// +unionDiscriminator` before a field means that this field is the
-  discriminator for the union. Only one field per structure can have this
-  prefix. This field HAS TO be a string, and CAN be optional.
-
-Multiple unions can exist per structure, but unions can't span across multiple
-go structures (all the fields that are part of a union has to be together in the
+  discriminator for the union. This field MUST be a string. Can be optional or
```

Comment:
Optional discriminator is confusing, I'd always make it required, though you can allow empty discriminator to mean unset.